examples/gameroom/daemon/src/rest_api/routes/authenticate.rs

```diff
@@ -32,28 +26,151 @@ pub struct AuthResponseData {
-        Err(err) => match err {
-            error::BlockingError::Error(_) => Ok(HttpResponse::Unauthorized()
-                .json(ErrorResponse::unauthorized("Invalid email or password"))),
-            error::BlockingError::Canceled => {
-                debug!("Internal Server Error: {}", err);
-                Ok(HttpResponse::InternalServerError().json(ErrorResponse::internal_error()))
+    // forward login to splinterd
+    let mut login_response = client
+        .post(format!("{}/biome/login", splinterd_url.get_ref()))
+        .header(
+            "SplinterProtocolVersion",
+            protocol::ADMIN_PROTOCOL_VERSION.to_string(),
+        )
+        .send_json(&UsernamePassword::from(auth_data.into_inner()))
+        .await?;
+
+    let token: Token = match login_response.status() {
+        StatusCode::OK => login_response.json().await?,
+        StatusCode::BAD_REQUEST => {
+            let body = login_response.body().await?;
+            let body_value: serde_json::Value = serde_json::from_slice(&body)?;
+            let message = match body_value.get("message") {
+                Some(value) => value.as_str().unwrap_or("Request was malformed."),
+                None => "Request malformed.",
+            };
+            return Ok(HttpResponse::BadRequest().json(ErrorResponse::bad_request(&message)));
+        }
+        StatusCode::UNAUTHORIZED => {
+            let body = login_response.body().await?;
+            let body_value: serde_json::Value = serde_json::from_slice(&body)?;
+            let message = match body_value.get("message") {
+                Some(value) => value.as_str().unwrap_or("Request was malformed."),
+                None => "Request malformed.",
+            };
+            return Ok(HttpResponse::Unauthorized().json(ErrorResponse::unauthorized(&message)));
+        }
+        _ => {
+            error!(
+                "Internal Server Error. Splinterd responded with error {}",
+                login_response.status(),
+            );
+            return Ok(HttpResponse::InternalServerError().json(ErrorResponse::internal_error()));
+        }
+    };
+
+    // Get user's key
+    let mut key_response = client
+        .get(format!(
+            "{}/biome/users/{}/keys",
+            splinterd_url.get_ref(),
+            token.user_id
+        ))
+        .header(
+            "SplinterProtocolVersion",
+            protocol::ADMIN_PROTOCOL_VERSION.to_string(),
+        )
+        .header("Authorization", token.token.clone())
+        .send()
+        .await?;
+
+    let key = match key_response.status() {
+        StatusCode::OK => {
+            let data = key_response.json::<Keys>().await?.data;
+            if let Some(key) = data.get(0) {
+                key.clone()
+            } else {
+                debug!("User does not have any keys");
+                return Ok(
+                    HttpResponse::InternalServerError().json(ErrorResponse::internal_error())
```

Comment:
This doesn't seem like an internal server error. Maybe the response should allow for the keys to be None?

Reply:
When a user registers via gameroom, part of that process is creating a key for that user via biome key api. So if on login no key is found for that user it would be an error.